test/unit/test_updates.coffee

```diff
@@ -247,122 +247,113 @@ QUnit.module "TaskList updates",
 
     @blockquote.append @field
 
+    @completeItemSourcePos = '1:1-1:14'
+    @incompleteItemSourcePos = '2:1-2:16'
+    @incompleteNBSPItemSourcePos = '3:1-3:27'
+    @quotedCompleteItemSourcePos = '4:3-4:23'
+    @quotedIncompleteItemSourcePos = '5:3-5:25'
+    @innerCompleteItemSourcePos = '6:4-6:23'
+    @innerIncompleteItemSourcePos = '7:5-7:26'
+    @orderedCompleteItemSourcePos = '8:3-8:25'
+    @orderedIncompleteItemSourcePos = '9:3-9:27'
+
     $('#qunit-fixture').append(@container)
     @container.taskList()
 
   afterEach: ->
     $(document).off 'tasklist:changed'
 
-QUnit.test "updates the source, marking the incomplete item as complete", (assert) ->
+assertTaskComplete =(assert, field, changes, taskItem, taskCheckbox, sourcepos) ->
   done = assert.async()
   assert.expect 3
 
-  @field.on 'tasklist:changed', (event) =>
+  taskItem.attr('data-sourcepos', sourcepos) if sourcepos
+
+  field.on 'tasklist:changed', (event) =>
     assert.ok event.detail.checked
-    assert.equal event.detail.index, @incompleteItem.expectedIndex
-    assert.equal @field.val(), @changes.toIncomplete
+    assert.equal event.detail.index, taskItem.expectedIndex
+    assert.equal field.val(), changes
     done()
 
-  @incompleteCheckbox.click()
+  taskCheckbox.click()
 
-QUnit.test "updates the source, marking the complete item as incomplete", (assert) ->
+assertTaskIncomplete =(assert, field, changes, taskItem, taskCheckbox, sourcepos) ->
   done = assert.async()
   assert.expect 3
 
-  @field.on 'tasklist:changed', (event) =>
+  taskItem.attr('data-sourcepos', sourcepos) if sourcepos
+
+  field.on 'tasklist:changed', (event) =>
     assert.ok !event.detail.checked
-    assert.equal event.detail.index, @completeItem.expectedIndex
-    assert.equal @field.val(), @changes.toComplete
+    assert.equal event.detail.index, taskItem.expectedIndex
+    assert.equal field.val(), changes
     done()
 
-  @completeCheckbox.click()
+  taskCheckbox.click()
+
+QUnit.test "updates the source, marking the incomplete item as complete", (assert) ->
+  assertTaskComplete(assert, @field, @changes.toIncomplete, @incompleteItem, @incompleteCheckbox, null)
```

Comment:
I think coupling this amount of testing logic into a single utility hurts the readability overall.

If you want to reduce the code duplication I would recommend doing it at a slightly more granular level. Maybe `@setSourcePosition(item, pos)`, `@on.changed.assertEventHas(prop, value)`, and `@on.changed.assertField(value)`?

That would make this test

```ts
QUnit.test "updates the source, marking the incomplete item as complete", (assert) ->
  @on.changed.assertEventHas('checked', true)
  @on.changed.assertEventHas('index', @incompleteItem.expectedIndex)
  @on.changed.assertField(@changes.toIncomplete)
  
  @incompleteCheckbox.click()
```

Feel free to decline this refactor, but I would prefer copying and pasting these test cases over the proposed test factories.

Reply:
No problem, I'll get it fixed up, but probably won't be until tomorrow...

Reply:
@deckar01 sorry, but my JS (and JS testing skills) are quite rusty.  I'm not quite sure how to build the `@onChanged` and `eventHas` functions for this...can you give me some direction?

Reply:
No worries. Closures can get a little cryptic in coffeescript.

```ts
@onChanged = (assert) =>
  utils =
  test: (fn) =>
    done = assert.async()
    @field.on 'tasklist:changed', (event) =>
      fn event
      done()
  eventHas: (name, value) =>
    utils.test (event) =>
      assert.equal event.detail[name], value
  fieldIs: (value) =>
    utils.test () =>
      assert.equal field.val(), value
```

Reply:
Thanks @deckar01 !  I've incorporated the changes, let me know what you think.

Reply:
Prefect. Thank you!